File: pepper_navigation/pepper_navigation/send_goal.py
"""
Entry point for a one-shot Nav2 navigation goal sender.

send_goal.py

Sets the robot's initial pose, waits for Nav2 to become active, and sends a
single navigation goal using the Nav2 Simple Commander API.

On startup, the script initializes a BasicNavigator, publishes an initial
pose at the map origin, waits until Nav2 is fully active, then sends a fixed
goal pose to the navigation stack via the NavigateToPose action. While the
goal is in progress, it periodically prints the remaining distance to the
goal until the task completes.

Actions:
    /navigate_to_pose (nav2_msgs/action/NavigateToPose)
        Action used (via BasicNavigator.goToPose) to drive the robot to the
        configured goal pose.

Author: Yohannes Tadesse Haile
Affiliation: Carnegie Mellon University Africa
Email: yohatad123@gmail.com
Date: March 08, 2026
Version: v1.0
"""

import rclpy
from geometry_msgs.msg import PoseStamped
from nav2_simple_commander.robot_navigator import BasicNavigator


def main():
    rclpy.init()
    navigator = BasicNavigator()

    # Set initial pose
    initial_pose = PoseStamped()
    initial_pose.header.frame_id = 'map'
    initial_pose.header.stamp = navigator.get_clock().now().to_msg()
    initial_pose.pose.position.x = 0.0
    initial_pose.pose.position.y = 0.0
    initial_pose.pose.orientation.w = 1.0
    navigator.setInitialPose(initial_pose)

    # Wait for Nav2 to activate
    navigator.waitUntilNav2Active()

    # Send goal
    goal_pose = PoseStamped()
    goal_pose.header.frame_id = 'map'
    goal_pose.header.stamp = navigator.get_clock().now().to_msg()
    goal_pose.pose.position.x = 2.0
    goal_pose.pose.position.y = 1.0
    goal_pose.pose.orientation.w = 1.0

    navigator.goToPose(goal_pose)

    while not navigator.isTaskComplete():
        feedback = navigator.getFeedback()
        print(f'Distance remaining: {feedback.distance_remaining}')

    print('Goal reached!')
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: pepper_navigation/src/tools/send_goal.cpp
/* send_goal.cpp
 *
 * One-shot Nav2 navigation goal sender. Publishes an initial pose at the
 * map origin, waits for the NavigateToPose action server to come up, then
 * sends a single fixed goal pose and prints the remaining distance until
 * the goal completes.
 *
 * nav2_simple_commander (the Python BasicNavigator this replaces) is a
 * Python-only convenience wrapper with no official C++ equivalent, so this
 * reimplements the relevant subset directly: a plain publish to
 * /initialpose (what AMCL listens to for manual pose initialization) and
 * an rclcpp_action client for NavigateToPose.
 *
 * Author: Yohannes Tadesse Haile
 * Affiliation: Carnegie Mellon University Africa
 * Email: yohatad123@gmail.com
 * Date: July 05, 2026
 * Version: v1.0 - C++ port of send_goal.py
 */

#include <chrono>
#include <memory>

#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>
#include <geometry_msgs/msg/pose_with_covariance_stamped.hpp>
#include <nav2_msgs/action/navigate_to_pose.hpp>

using NavigateToPose = nav2_msgs::action::NavigateToPose;
using GoalHandle = rclcpp_action::ClientGoalHandle<NavigateToPose>;

class SendGoal : public rclcpp::Node {
public:
    SendGoal() : Node("send_goal") {
        initial_pose_pub_ = create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>(
            "/initialpose", 10);
        action_client_ = rclcpp_action::create_client<NavigateToPose>(this, "navigate_to_pose");
    }

    void run() {
        publishInitialPose();

        RCLCPP_INFO(get_logger(), "Waiting for navigate_to_pose action server...");
        if (!action_client_->wait_for_action_server(std::chrono::seconds(30))) {
            RCLCPP_ERROR(get_logger(), "navigate_to_pose action server not available");
            return;
        }

        NavigateToPose::Goal goal;
        goal.pose.header.frame_id = "map";
        goal.pose.header.stamp = now();
        goal.pose.pose.position.x = 2.0;
        goal.pose.pose.position.y = 1.0;
        goal.pose.pose.orientation.w = 1.0;

        auto send_goal_options = rclcpp_action::Client<NavigateToPose>::SendGoalOptions();
        send_goal_options.feedback_callback =
            [this](GoalHandle::SharedPtr, const std::shared_ptr<const NavigateToPose::Feedback> feedback) {
                RCLCPP_INFO(get_logger(), "Distance remaining: %.3f", feedback->distance_remaining);
            };

        auto goal_handle_future = action_client_->async_send_goal(goal, send_goal_options);
        if (rclcpp::spin_until_future_complete(get_node_base_interface(), goal_handle_future) !=
            rclcpp::FutureReturnCode::SUCCESS) {
            RCLCPP_ERROR(get_logger(), "Failed to send goal");
            return;
        }
        auto goal_handle = goal_handle_future.get();
        if (!goal_handle) {
            RCLCPP_ERROR(get_logger(), "Goal was rejected");
            return;
        }

        auto result_future = action_client_->async_get_result(goal_handle);
        if (rclcpp::spin_until_future_complete(get_node_base_interface(), result_future) !=
            rclcpp::FutureReturnCode::SUCCESS) {
            RCLCPP_ERROR(get_logger(), "Failed to get result");
            return;
        }
        RCLCPP_INFO(get_logger(), "Goal reached!");
    }

private:
    void publishInitialPose() {
        geometry_msgs::msg::PoseWithCovarianceStamped initial_pose;
        initial_pose.header.frame_id = "map";
        initial_pose.header.stamp = now();
        initial_pose.pose.pose.position.x = 0.0;
        initial_pose.pose.pose.position.y = 0.0;
        initial_pose.pose.pose.orientation.w = 1.0;
        initial_pose_pub_->publish(initial_pose);
    }

    rclcpp::Publisher<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr initial_pose_pub_;
    rclcpp_action::Client<NavigateToPose>::SharedPtr action_client_;
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SendGoal>();
    node->run();
    rclcpp::shutdown();
    return 0;
}
